docs/en/sql-reference/functions/json-functions.md

```diff
@@ -1155,3 +1155,207 @@ SELECT jsonMergePatch('{"a":1}', '{"name": "joey"}', '{"name": "tom"}', '{"name"
+JSONAllPathsWithTypes(json)
+```
+
+**Arguments**
+
+- `json` — [JSON](../data-types/newjson.md).
+
+**Returned value**
+
+- An array of paths. [Map(String, String)](../data-types/array.md).
+
+**Example**
+
+``` sql
+CREATE TABLE test (json JSON(max_dynamic_paths=1)) ENGINE = Memory;
+INSERT INTO test FORMAT JSONEachRow {"json" : {"a" : 42}}, {"json" : {"b" : "Hello"}}, {"json" : {"a" : [1, 2, 3], "c" : "2020-01-01"}}
+SELECT json, JSONDynamicPathsWithTypes(json) FROM test;
+```
+
+```text
+┌─json─────────────────────────────────┬─JSONDynamicPathsWithTypes(json)─┐
+│ {"a":"42"}                           │ {'a':'Int64'}                   │
+│ {"b":"Hello"}                        │ {}                              │
+│ {"a":["1","2","3"],"c":"2020-01-01"} │ {'a':'Array(Nullable(Int64))'}  │
+└──────────────────────────────────────┴─────────────────────────────────┘
+```
+
+### JSONSharedDataPaths
+
+Returns the list of paths that are stored in shared data structure in [JSON](../data-types/newjson.md) column.
+
+**Syntax**
+
+``` sql
+JSONSharedDataPaths(json)
+```
+
+**Arguments**
+
+- `json` — [JSON](../data-types/newjson.md).
+
+**Returned value**
+
+- An array of paths. [Array(String)](../data-types/array.md).
+
+**Example**
+
+``` sql
+CREATE TABLE test (json JSON(max_dynamic_paths=1)) ENGINE = Memory;
+INSERT INTO test FORMAT JSONEachRow {"json" : {"a" : 42}}, {"json" : {"b" : "Hello"}}, {"json" : {"a" : [1, 2, 3], "c" : "2020-01-01"}}
+SELECT json, JSONSharedDataPaths(json) FROM test;
+```
+
+```text
+┌─json─────────────────────────────────┬─JSONSharedDataPaths(json)─┐
+│ {"a":"42"}                           │ []                        │
+│ {"b":"Hello"}                        │ ['b']                     │
+│ {"a":["1","2","3"],"c":"2020-01-01"} │ ['c']                     │
+└──────────────────────────────────────┴───────────────────────────┘
+```
+
+### JSONSharedDataPathsWithTypes
+
+Returns the map of paths that are stored in shared data structure and their types in each row in [JSON](../data-types/newjson.md) column.
+
+**Syntax**
+
+``` sql
+JSONSharedDataPathsWithTypes(json)
+```
+
+**Arguments**
+
+- `json` — [JSON](../data-types/newjson.md).
+
+**Returned value**
+
+- An array of paths. [Map(String, String)](../data-types/array.md).
+
+**Example**
+
+``` sql
+CREATE TABLE test (json JSON(max_dynamic_paths=1)) ENGINE = Memory;
+INSERT INTO test FORMAT JSONEachRow {"json" : {"a" : 42}}, {"json" : {"b" : "Hello"}}, {"json" : {"a" : [1, 2, 3], "c" : "2020-01-01"}}
+SELECT json, JSONSharedDataPathsWithTypes(json) FROM test;
+```
+
+```text
+┌─json─────────────────────────────────┬─JSONSharedDataPathsWithTypes(json)─┐
+│ {"a":"42"}                           │ {}                                 │
+│ {"b":"Hello"}                        │ {'b':'String'}                     │
+│ {"a":["1","2","3"],"c":"2020-01-01"} │ {'c':'Date'}                       │
+└──────────────────────────────────────┴────────────────────────────────────┘
+```
```

Comment:
I am trying to understand this. If I changed the insert into:

```sql
INSERT INTO test FORMAT JSONEachRow {"json" : {"a" : 42}}, {"json" : {"b" : [{"v":"h","w":1}]}}, {"json" : {"a" : [1, 2, 3], "c" : "2020-01-01"}}
```
Then `JSONSharedDataPathsWithTypes` will return `{}` for every row. Is this expected?